chain/index/events.go

```diff
@@ -0,0 +1,548 @@
+					}
+				} else {
+					ce.EmitterAddr, err = address.NewFromBytes(row.emitterAddr)
+					if err != nil {
+						return nil, xerrors.Errorf("parse emitter addr: %w", err)
+					}
+				}
+
+				tsKeyCid, err := cid.Cast(row.tipsetKeyCid)
+				if err != nil {
+					return nil, xerrors.Errorf("parse tipsetkey cid: %w", err)
+				}
+
+				ts, err := si.cs.GetTipSetByCid(ctx, tsKeyCid)
+				if err != nil {
+					return nil, xerrors.Errorf("get tipset by cid: %w", err)
+				}
+
+				ce.TipSetKey = ts.Key()
+
+				ce.MsgCid, err = cid.Cast(row.messageCid)
+				if err != nil {
+					return nil, xerrors.Errorf("parse message cid: %w", err)
+				}
+			}
+
+			ce.Entries = append(ce.Entries, types.EventEntry{
+				Flags: row.flags[0],
+				Key:   row.key,
+				Codec: row.codec,
+				Value: row.value,
+			})
+		}
+
+		if ce != nil {
+			ces = append(ces, ce)
+		}
+
+		if len(ces) == 0 {
+			return nil, nil
+		}
+
+		// collected event list is in inverted order since we selected only the most recent events
+		// sort it into height order
+		sort.Slice(ces, func(i, j int) bool { return ces[i].Height < ces[j].Height })
+
+		return ces, nil
+	}
+
+	values, query, err := makePrefillFilterQuery(f, excludeReverted)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to make prefill filter query: %w", err)
+	}
+
+	stmt, err := si.db.Prepare(query)
+	if err != nil {
+		return nil, xerrors.Errorf("prepare prefill query: %w", err)
+	}
+	defer func() { _ = stmt.Close() }()
+
+	ces, err := getEventsFnc(stmt, values)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get events: %w", err)
+	}
+	if len(ces) == 0 {
+		// there's no matching events for the filter, wait till index has caught up to the head and then retry
+		if err := si.waitTillHeadIndexed(ctx); err != nil {
+			return nil, xerrors.Errorf("failed to wait for head to be indexed: %w", err)
+		}
+		ces, err = getEventsFnc(stmt, values)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to get events: %w", err)
+		}
+
+		if len(ces) == 0 {
+			return nil, si.checkTipsetIndexedStatus(ctx, f)
+		}
+	}
+
+	return ces, nil
+}
+
+func makePrefillFilterQuery(f *EventFilter, excludeReverted bool) ([]any, string, error) {
+	clauses := []string{}
+	values := []any{}
+	joins := []string{}
+
+	if f.TipsetCid != cid.Undef {
+		clauses = append(clauses, "tm.tipset_key_cid=?")
+		values = append(values, f.TipsetCid.Bytes())
+	} else {
+		if f.MinHeight >= 0 && f.MinHeight == f.MaxHeight {
+			clauses = append(clauses, "tm.height=?")
+			values = append(values, f.MinHeight)
+		} else {
+			if f.MaxHeight >= 0 && f.MinHeight >= 0 {
+				clauses = append(clauses, "tm.height BETWEEN ? AND ?")
+				values = append(values, f.MinHeight, f.MaxHeight)
+			} else if f.MinHeight >= 0 {
+				clauses = append(clauses, "tm.height >= ?")
+				values = append(values, f.MinHeight)
+			} else if f.MaxHeight >= 0 {
+				clauses = append(clauses, "tm.height <= ?")
+				values = append(values, f.MaxHeight)
+			}
```

Comment:
```suggestion
			} else {
				return nil, "", xerrors.Errorf("filter must specify either a tipset or a height range")
			}
```

we're missing this from the current form, and assuming that the caller has done this check, but we should probably do it ourselves!

Reply:
You'd better validate my reasoning here - there should never be a call to this that doesn't specify either a tipset or height range, correct? If there is then we may have trouble with query times because we use those to narrow down our selections.

Reply:
@rvagg Yeah one of those params is mandatory.